### PR TITLE
Implements new indicator IBS - Internal Bar Strength

### LINKED ISSUE #8124: Implement Internal Bar Strenght
#### Expected Behavior
Supports Internal Bar Strength (see [talipp implementation](https://github.com/nardew/talipp/blob/main/talipp/indicators/IBS.py)).
We use this indicator in the `GlobalEquityMeanReversionIBSAlpha`, so we should use the new indicator there. 

Refer to [LEAN Engine > Contributions > Indicators](https://www.quantconnect.com/docs/v2/lean-engine/contributions/indicators).

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by femtotrader**
#### Description

Implement new indicator - IBS

#### Related Issue

Closes #8124

#### Motivation and Context
The InternalBarStrenght indicator is a measure of the relative position of a period's closing price to the same period's high and low.

#### Requires Documentation Change

Doc for this indicator should be created but some docstring are written

#### How Has This Been Tested?

comparison with talipp computation of IBS() of SPY candle data using [generate_reference_data_from_talipp.py](https://github.com/QuantConnect/Lean/blob/master/Tests/TestData/generate_reference_data_from_talipp.py)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice! Thank you, 
Leaving some minor requests, needs a rebase from master too sorry 👍

File: Indicators/InternalBarStrength.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            if (input.High == input.Low)
+            {
+                if (input.Close > input.Low)
+                {
+                    return decimal.MaxValue;
+                }
+                else
+                {
+                    return decimal.MinValue;
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/InternalBarStrength.cs:70]: Maybe they are simply raising an exception.

If we don't want to raise an exception, than [sentinel values](https://en.wikipedia.org/wiki/Sentinel_value) are a solution

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Indicators/InternalBarStrength.cs:70]: > **Maybe** they are simply raising an exception.

Right, should review how other platforms handle this case, which value to they assign to the indicator? Please share links/images/proof

> If we don't want to raise an exception, than [sentinel values](https://en.wikipedia.org/wiki/Sentinel_value) are a solution

Right, exception is bad, like the sentinel value 👍 which value though? Imagine user plotting the indicator in a chart, using max/min will make it worthless

File: Indicators/InternalBarStrength.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+using QuantConnect.Data.Market;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// The InternalBarStrenght indicator is a measure of the relative position of a period's closing price 
+    /// to the same period's high and low.
+    //  The IBS can be interpreted to predict a bullish signal when displaying a low value and a bearish signal when presenting a high value.
+    /// </summary>
+    public class InternalBarStrength : TradeBarIndicator, IIndicatorWarmUpPeriodProvider
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Indicators/InternalBarStrength.cs:25]: Rebase @femtotrader 
```
## REBASE origin/master
git checkout master
git fetch upstream
git rebase upstream/master
git push
```
next
```
## REBASE BRANCH FROM MASTER
git checkout {my_branch}
git rebase master
git push --force-with-lease
```

File: Indicators/InternalBarStrength.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/InternalBarStrength.cs:70]: talipp let the exception occurs

https://github.com/nardew/talipp/blob/ed5e0e10c986121b5762d4538649212ec05a9847/talipp/indicators/IBS.py#L38

freqtrade
https://github.com/freqtrade/freqtrade/blob/dcedc1c6525c5f767951fe93335a01cb2f83ed5c/freqtrade/vendor/qtpylib/indicators.py#L192
use Pandas (with float... so we have NaN)

Zorro Trader return 1.
https://github.com/AttilaForgacs/zorro/blob/2c851f35c004cceb761fdc97c2a93fe82f6d399e/strat/include/indicators.c#L281

File: Indicators/InternalBarStrength.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/InternalBarStrength.cs:25]: I pushed in feature branch without (--force-with-lease) option (that I didn't know)

I now have 38 files in this PR ! 😭 

My gitfu is a mess

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Indicators/InternalBarStrength.cs]: yes sorry, seems u'll need to squash the history or create a new PR 👍

[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/InternalBarStrength.cs]: Looks good now... only having 5 files in this PR now (but with 16 commits !)
Squashing is a nightmare. Maybe you can do on your side when merging.

File: Indicators/InternalBarStrength.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Indicators/InternalBarStrength.cs]: Hey @femtotrader! Mind addressing this point first, so we can ship this PR before working on newer ones 🙏

[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/InternalBarStrength.cs]: In this PR we are returning 1 (decimal) in such a case as Zorro Trader is doing.


https://github.com/QuantConnect/Lean/pull/8151/files#diff-954d02ea08ad7f931e33b5677e14dc20aa8f50b56f2e9b3a4dc2afb38148072cR64

Chart won't help.

But it's important to say that 0 / 0 is an [indeterminate_form](https://en.wikipedia.org/wiki/Indeterminate_form) so any value have some "sense".

I asked Claude.ai about it.

Here is his "point of view"

> please explain what IBS Internal Bar Strength (technical analysis indicator) is

Internal Bar Strength (IBS) is a technical indicator used by traders to measure the closing price of a security relative to its trading range for a specific period. Here are the key points about this IBS indicator:

1. Calculation: IBS = (Close - Low) / (High - Low)
Where Close is the closing price, Low is the lowest price, and High is the highest price for the period.

2. Range: IBS values always fall between 0 and 1.
Interpretation:

- IBS = 0: The closing price equals the low of the period.
- IBS = 1: The closing price equals the high of the period.
- IBS = 0.5: The closing price is exactly in the middle of the range.

4. Usage: Traders use IBS to gauge the strength or weakness of price action within a specific time frame.
5. Trend indication:
- Higher IBS values suggest bullish sentiment.
- Lower IBS values suggest bearish sentiment.

6. Divergence: Traders often look for divergences between IBS and price to identify potential trend reversals.
7. Timeframes: IBS can be applied to various timeframes, from intraday charts to weekly or monthly periods.
8. Complementary tools: IBS is often used in conjunction with other technical indicators for more comprehensive analysis.

> What is it value when High = Low

When the High price equals the Low price, we encounter a special case for the Internal Bar Strength (IBS) indicator. Let's examine this situation:
The IBS formula is: IBS = (Close - Low) / (High - Low)
When High = Low, we have a situation where:

1. The price didn't move during the period (a doji candle in candlestick charting).
2. The denominator of the formula becomes zero: (High - Low) = 0

In mathematics, division by zero is undefined. Therefore, the IBS value is technically undefined in this scenario.
In practice, how this is handled depends on the specific implementation of the indicator in trading software:

1. Some systems might return an error or a null value.
2. Others might default to a specific value, such as 0.5 (assuming the close is at the middle of the non-existent range).
3. Some implementations might use the previous valid IBS value.

It's important for traders using the IBS indicator to be aware of how their particular trading platform or software handles this edge case. They should also consider the broader market context when encountering such situations, as a period with no price movement (High = Low) often indicates very low volatility or possibly a trading halt

...

Maybe 0.5 would be a more appropriate value?

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Indicators/InternalBarStrength.cs]: Sweet nice, 0.5 sounds great 🚀

[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/InternalBarStrength.cs]: When High = Low, 0.5 is now returned thanks to https://github.com/QuantConnect/Lean/pull/8151/commits/5744cc639d43ba36d98ed730eb89d3fee11f08ff

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**